File: app/src/main/java/io/nekohasekai/sagernet/auth/AuthRepository.kt
package io.nekohasekai.sagernet.auth

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.google.gson.Gson
import com.google.gson.JsonObject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * 认证仓库 - 处理所有认证相关的 API 调用
 */
class AuthRepository(private val context: Context) {
    
    private val TAG = "AuthRepository"
    private val BASE_URL = "https://dy.moneyfly.top"
    private val API_BASE = "$BASE_URL/api/v1"
    private val gson = Gson()
    
    private val prefs: SharedPreferences = context.getSharedPreferences(
        "auth_prefs",
        Context.MODE_PRIVATE
    )
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()
    
    /**
     * 登录
     */
    suspend fun login(email: String, password: String): Result<LoginResponse> = withContext(Dispatchers.IO) {
        try {
            val json = JsonObject().apply {
                addProperty("email", email)
                addProperty("password", password)
            }
            
            val request = Request.Builder()
                .url("$API_BASE/auth/login")
                .post(json.toString().toRequestBody("application/json".toMediaType()))
                .build()
            
            val response = client.newCall(request).execute()
            val body = response.body?.string()
            
            Log.d(TAG, "登录请求: ${request.url}")
            Log.d(TAG, "登录响应码: ${response.code}")
            Log.d(TAG, "登录响应体: $body")
            
            if (response.isSuccessful && body != null) {
                val apiResponse = gson.fromJson(body, JsonObject::class.java)
                
                if (apiResponse.get("success")?.asBoolean == true) {
                    val data = apiResponse.getAsJsonObject("data")
                    val user = data.getAsJsonObject("user")
                    val loginResponse = LoginResponse(
                        token = data.get("access_token").asString,
                        email = user.get("email").asString,
                        username = user.get("username")?.asString ?: user.get("email").asString
                    )
                    
                    // 保存 token
                    saveToken(loginResponse.token)
                    saveEmail(loginResponse.email)
                    saveUsername(loginResponse.username)
                    
                    Log.d(TAG, "登录成功: ${loginResponse.email}")
                    Result.success(loginResponse)
                } else {
                    val message = apiResponse.get("message")?.asString ?: "登录失败"
                    Log.e(TAG, "登录失败: $message")
                    Result.failure(Exception(message))
                }
            } else {
                val errorMessage = try {
                    if (body != null) {
                        val apiResponse = gson.fromJson(body, JsonObject::class.java)
                        apiResponse.get("message")?.asString ?: "登录失败，请检查网络连接"
                    } else {
                        "登录失败，请检查网络连接"
                    }
                } catch (e: Exception) {
                    "登录失败，请检查网络连接"
                }
                Log.e(TAG, "登录请求失败: ${response.code}, 错误: $errorMessage")
                Result.failure(Exception(errorMessage))
            }
        } catch (e: Exception) {
            Log.e(TAG, "登录异常: ${e.message}", e)
            val errorMsg = when {
                e.message?.contains("timeout") == true -> "连接超时，请检查网络连接"
                e.message?.contains("SSL") == true -> "SSL 连接错误，请检查网络设置"
                e.message?.contains("failed to connect") == true -> "无法连接到服务器，请检查网络"
                else -> "登录失败: ${e.message ?: "未知错误"}"
            }
            Result.failure(Exception(errorMsg))
        }
    }
    
    /**
     * 注册
     */
    suspend fun register(
        username: String,
        email: String,
        password: String,
        verificationCode: String? = null,
        inviteCode: String? = null
    ): Result<String> = withContext(Dispatchers.IO) {
        try {
            val json = JsonObject().apply {
                addProperty("username", username)
                addProperty("email", email)
                addProperty("password", password)
                verificationCode?.let { addProperty("verification_code", it) }
                inviteCode?.let { addProperty("invite_code", it) }
            }
            
            val request = Request.Builder()
                .url("$API_BASE/auth/register")
                .post(json.toString().toRequestBody("application/json".toMediaType()))
                .build()
            
            val response = client.newCall(request).execute()
            val body = response.body?.string()
            
            if (response.isSuccessful && body != null) {
                val apiResponse = gson.fromJson(body, JsonObject::class.java)
                
                if (apiResponse.get("success")?.asBoolean == true) {
                    val message = apiResponse.get("message")?.asString ?: "注册成功"
                    Log.d(TAG, "注册成功: $email")
                    Result.success(message)
                } else {
                    val message = apiResponse.get("message")?.asString ?: "注册失败"
                    Log.e(TAG, "注册失败: $message")
                    Result.failure(Exception(message))
                }
            } else {
                Log.e(TAG, "注册请求失败: ${response.code}")
                Result.failure(Exception("注册失败，请检查网络连接"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "注册异常", e)
            Result.failure(e)
        }
    }
    
    /**
     * 发送验证码（用于注册）
     */
    suspend fun sendVerificationCode(email: String, type: String = "register"): Result<String> = withContext(Dispatchers.IO) {
        try {
            val json = JsonObject().apply {
                addProperty("email", email)
                addProperty("type", "email") // 后端要求 type 为 "email"
            }
            
            val request = Request.Builder()
                .url("$API_BASE/auth/verification/send")
                .post(json.toString().toRequestBody("application/json".toMediaType()))
                .build()
            
            val response = client.newCall(request).execute()
            val body = response.body?.string()
            
            Log.d(TAG, "验证码请求: ${request.url}")
            Log.d(TAG, "验证码响应码: ${response.code}")
            Log.d(TAG, "验证码响应体: $body")
            
            if (response.isSuccessful && body != null) {
                val apiResponse = gson.fromJson(body, JsonObject::class.java)
                
                if (apiResponse.get("success")?.asBoolean == true) {
                    val message = apiResponse.get("message")?.asString ?: "验证码已发送"
                    Log.d(TAG, "验证码发送成功: $email")
                    Result.success(message)
                } else {
                    val message = apiResponse.get("message")?.asString ?: "发送失败"
                    Log.e(TAG, "验证码发送失败: $message")
                    Result.failure(Exception(message))
                }
            } else {
                val errorMessage = try {
                    if (body != null) {
                        val apiResponse = gson.fromJson(body, JsonObject::class.java)
                        apiResponse.get("message")?.asString ?: "发送失败，请检查网络连接"
                    } else {
                        "发送失败，请检查网络连接"
                    }
                } catch (e: Exception) {
                    "发送失败，请检查网络连接"
                }
                Log.e(TAG, "验证码请求失败: ${response.code}, 错误: $errorMessage")
                Result.failure(Exception(errorMessage))
            }
        } catch (e: Exception) {
            Log.e(TAG, "验证码发送异常: ${e.message}", e)
            val errorMsg = when {
                e.message?.contains("timeout") == true -> "连接超时，请检查网络连接"
                e.message?.contains("SSL") == true -> "SSL 连接错误，请检查网络设置"
                else -> "发送失败: ${e.message ?: "未知错误"}"
            }
            Result.failure(Exception(errorMsg))
        }
    }
    
    /**
     * 忘记密码（发送重置密码验证码）
     */
    suspend fun forgotPassword(email: String): Result<String> = withContext(Dispatchers.IO) {
        try {
            val json = JsonObject().apply {
                addProperty("email", email)
            }
            
            val request = Request.Builder()
                .url("$API_BASE/auth/forgot-password")
                .post(json.toString().toRequestBody("application/json".toMediaType()))
                .build()
            
            val response = client.newCall(request).execute()
            val body = response.body?.string()
            
            Log.d(TAG, "忘记密码请求: ${request.url}")
            Log.d(TAG, "忘记密码响应码: ${response.code}")
            Log.d(TAG, "忘记密码响应体: $body")
            
            if (response.isSuccessful && body != null) {
                val apiResponse = gson.fromJson(body, JsonObject::class.java)
                
                if (apiResponse.get("success")?.asBoolean == true) {
                    val message = apiResponse.get("message")?.asString ?: "验证码已发送"
                    Log.d(TAG, "忘记密码验证码发送成功: $email")
                    Result.success(message)
                } else {
                    val message = apiResponse.get("message")?.asString ?: "发送失败"
                    Log.e(TAG, "忘记密码验证码发送失败: $message")
                    Result.failure(Exception(message))
                }
            } else {
                val errorMessage = try {
                    if (body != null) {
                        val apiResponse = gson.fromJson(body, JsonObject::class.java)
                        apiResponse.get("message")?.asString ?: "发送失败，请检查网络连接"
                    } else {
                        "发送失败，请检查网络连接"
                    }
                } catch (e: Exception) {
                    "发送失败，请检查网络连接"
                }
                Log.e(TAG, "忘记密码请求失败: ${response.code}, 错误: $errorMessage")
                Result.failure(Exception(errorMessage))
            }
        } catch (e: Exception) {
            Log.e(TAG, "忘记密码异常: ${e.message}", e)
            val errorMsg = when {
                e.message?.contains("timeout") == true -> "连接超时，请检查网络连接"
                e.message?.contains("SSL") == true -> "SSL 连接错误，请检查网络设置"
                else -> "发送失败: ${e.message ?: "未知错误"}"
            }
            Result.failure(Exception(errorMsg))
        }
    }
    
    /**
     * 重置密码
     */
    suspend fun resetPassword(email: String, code: String, newPassword: String): Result<String> = withContext(Dispatchers.IO) {
        try {
            val json = JsonObject().apply {
                addProperty("email", email)
                addProperty("verification_code", code)
                addProperty("new_password", newPassword)
            }
            
            val request = Request.Builder()
                .url("$API_BASE/auth/reset-password")
                .post(json.toString().toRequestBody("application/json".toMediaType()))
                .build()
            
            val response = client.newCall(request).execute()
            val body = response.body?.string()
            
            if (response.isSuccessful && body != null) {
                val apiResponse = gson.fromJson(body, JsonObject::class.java)
                
                if (apiResponse.get("success")?.asBoolean == true) {
                    val message = apiResponse.get("message")?.asString ?: "密码重置成功"
                    Log.d(TAG, "密码重置成功: $email")
                    Result.success(message)
                } else {
                    val message = apiResponse.get("message")?.asString ?: "重置失败"
                    Log.e(TAG, "密码重置失败: $message")
                    Result.failure(Exception(message))
                }
            } else {
                Log.e(TAG, "重置密码请求失败: ${response.code}")
                Result.failure(Exception("重置失败，请检查网络连接"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "重置密码异常", e)
            Result.failure(e)
        }
    }
    
    /**
     * 刷新 Token
     */
    suspend fun refreshToken(): Result<String> = withContext(Dispatchers.IO) {
        try {
            val currentToken = getToken()
            if (currentToken == null) {
                return@withContext Result.failure(Exception("无 Token"))
            }
            
            val request = Request.Builder()
                .url("$API_BASE/auth/refresh")
                .header("Authorization", "Bearer $currentToken")
                .post("".toRequestBody())
                .build()
            
            val response = client.newCall(request).execute()
            val body = response.body?.string()
            
            if (response.isSuccessful && body != null) {
                val apiResponse = gson.fromJson(body, JsonObject::class.java)
                
                if (apiResponse.get("success")?.asBoolean == true) {
                    val data = apiResponse.getAsJsonObject("data")
                    val newToken = data.get("token").asString
                    
                    // 保存新 Token
                    saveToken(newToken)
                    
                    Log.d(TAG, "Token 刷新成功")
                    Result.success(newToken)
                } else {
                    val message = apiResponse.get("message")?.asString ?: "Token 刷新失败"
                    Log.e(TAG, "Token 刷新失败: $message")
                    Result.failure(Exception(message))
                }
            } else if (response.code == 401) {
                // Token 已完全失效，需要重新登录
                Log.e(TAG, "Token 已失效，需要重新登录")
                logout()
                Result.failure(Exception("登录已过期，请重新登录"))
            } else {
                Log.e(TAG, "Token 刷新请求失败: ${response.code}")
                Result.failure(Exception("Token 刷新失败"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Token 刷新异常", e)
            Result.failure(e)
        }
    }
    
    /**
     * 自动刷新 Token（带重试机制）
     * 在 API 调用失败时自动调用
     */
    suspend fun autoRefreshTokenIfNeeded(errorCode: Int): Boolean {
        if (errorCode == 401) {
            // Token 可能过期，尝试刷新
            val result = refreshToken()
            return result.isSuccess
        }
        return false
    }
    
    /**
     * 获取用户订阅
     */
    suspend fun getUserSubscription(): Result<UserSubscription> = withContext(Dispatchers.IO) {
        try {
            val token = getToken()
            if (token == null) {
                Log.e(TAG, "未登录，无法获取订阅")
                return@withContext Result.failure(Exception("未登录"))
            }
            
            val request = Request.Builder()
                .url("$API_BASE/subscriptions/user-subscription")
                .header("Authorization", "Bearer $token")
                .get()
                .build()
            
            val response = client.newCall(request).execute()
            val body = response.body?.string()
            
            if (response.isSuccessful && body != null) {
                val apiResponse = gson.fromJson(body, JsonObject::class.java)
                
                if (apiResponse.get("success")?.asBoolean == true) {
                    val data = apiResponse.getAsJsonObject("data")
                    val subscription = UserSubscription(
                        universalUrl = data.get("universal_url")?.asString ?: "",
                        subscriptionUrl = data.get("subscription_url")?.asString,
                        expireTime = data.get("expire_time")?.asString ?: "未设置",
                        deviceLimit = data.get("device_limit")?.asInt ?: 0,
                        currentDevices = data.get("current_devices")?.asInt ?: 0,
                        uploadTraffic = data.get("upload_traffic")?.asLong ?: 0,
                        downloadTraffic = data.get("download_traffic")?.asLong ?: 0,
                        totalTraffic = data.get("total_traffic")?.asLong ?: 0
                    )
                    
                    Log.d(TAG, "获取订阅成功: ${subscription.expireTime}")
                    Result.success(subscription)
                } else {
                    val message = apiResponse.get("message")?.asString ?: "获取订阅失败"
                    Log.e(TAG, "获取订阅失败: $message")
                    Result.failure(Exception(message))
                }
            } else {
                Log.e(TAG, "订阅请求失败: ${response.code}")
                Result.failure(Exception("获取订阅失败"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "获取订阅异常", e)
            Result.failure(e)
        }
    }
    
    // Token 管理
    fun saveToken(token: String) {
        prefs.edit().putString("auth_token", token).apply()
    }
    
    fun getToken(): String? {
        return prefs.getString("auth_token", null)
    }
    
    fun clearToken() {
        prefs.edit().remove("auth_token").apply()
    }
    
    fun saveEmail(email: String) {
        prefs.edit().putString("user_email", email).apply()
    }
    
    fun getEmail(): String? {
        return prefs.getString("user_email", null)
    }
    
    fun saveUsername(username: String) {
        prefs.edit().putString("user_username", username).apply()
    }
    
    fun getUsername(): String? {
        return prefs.getString("user_username", null)
    }
    
    fun isAuthenticated(): Boolean {
        return getToken() != null
    }
    
    fun logout() {
        prefs.edit().clear().apply()
        Log.d(TAG, "用户已退出登录")
    }
    
    fun getAuthState(): AuthState {
        return AuthState(
            isAuthenticated = isAuthenticated(),
            token = getToken(),
            email = getEmail(),
            username = getUsername()
        )
    }
}

File: windows_app/lib/features/auth/repositories/auth_repository.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../models/auth_models.dart';
import '../../../core/utils/logger.dart';

class AuthRepository {
  static const String _baseUrl = 'https://dy.moneyfly.top';
  static const String _apiBase = '$_baseUrl/api/v1';
  static const String _tokenKey = 'auth_token';
  static const String _emailKey = 'auth_email';
  static const String _usernameKey = 'auth_username';
  static const String _subscriptionUrlKey = 'subscription_url';
  static const String _expireTimeKey = 'expire_time';
  static const String _hasSubscriptionKey = 'has_subscription';
  static const String _maxDevicesKey = 'max_devices';
  static const String _onlineDevicesKey = 'online_devices';

  final http.Client _client = http.Client();

  // 获取保存的 token
  Future<String?> getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_tokenKey);
  }

  // 保存 token
  Future<void> saveToken(String token) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_tokenKey, token);
  }

  // 保存用户信息
  Future<void> saveUserInfo(String email, String username) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_emailKey, email);
    await prefs.setString(_usernameKey, username);
  }

  // 获取用户信息
  Future<Map<String, String?>> getUserInfo() async {
    final prefs = await SharedPreferences.getInstance();
    return {
      'email': prefs.getString(_emailKey),
      'username': prefs.getString(_usernameKey),
    };
  }

  // 检查是否已登录
  Future<bool> isAuthenticated() async {
    final token = await getToken();
    return token != null && token.isNotEmpty;
  }

  // 登录
  Future<Result<LoginResponse>> login(String email, String password) async {
    try {
      final response = await _client.post(
        Uri.parse('$_apiBase/auth/login'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'email': email,
          'password': password,
        }),
      ).timeout(const Duration(seconds: 30));

      Logger.debug('登录请求: $_apiBase/auth/login');
      Logger.debug('登录响应码: ${response.statusCode}');
      Logger.debug('登录响应体: ${response.body}');

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        if (json['success'] == true) {
          final loginResponse = LoginResponse.fromJson(json);
          await saveToken(loginResponse.token);
          await saveUserInfo(loginResponse.email, loginResponse.username);
          Logger.debug('登录成功: ${loginResponse.email}');
          return Result.success(loginResponse);
        } else {
          final message = json['message'] as String? ?? '登录失败';
          Logger.error('登录失败: $message');
          return Result.failure(Exception(message));
        }
      } else {
        final errorMessage = _parseErrorMessage(response.body);
        Logger.error('登录请求失败: ${response.statusCode}, 错误: $errorMessage');
        return Result.failure(Exception(errorMessage));
      }
    } catch (e) {
      Logger.error('登录异常: $e');
      final errorMsg = _getErrorMessage(e);
      return Result.failure(Exception(errorMsg));
    }
  }

  // 注册
  Future<Result<String>> register(
    String username,
    String email,
    String password, {
    String? verificationCode,
    String? inviteCode,
  }) async {
    try {
      final response = await _client.post(
        Uri.parse('$_apiBase/auth/register'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'username': username,
          'email': email,
          'password': password,
          if (verificationCode != null) 'verification_code': verificationCode,
          if (inviteCode != null) 'invite_code': inviteCode,
        }),
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        if (json['success'] == true) {
          final message = json['message'] as String? ?? '注册成功';
          Logger.debug('注册成功: $email');
          return Result.success(message);
        } else {
          final message = json['message'] as String? ?? '注册失败';
          Logger.error('注册失败: $message');
          return Result.failure(Exception(message));
        }
      } else {
        final errorMessage = _parseErrorMessage(response.body);
        Logger.error('注册请求失败: ${response.statusCode}, 错误: $errorMessage');
        return Result.failure(Exception(errorMessage));
      }
    } catch (e) {
      Logger.error('注册异常: $e');
      final errorMsg = _getErrorMessage(e);
      return Result.failure(Exception(errorMsg));
    }
  }

  // 发送验证码
  Future<Result<String>> sendVerificationCode(
    String email,
    String type,
  ) async {
    try {
      final token = await getToken();
      final headers = {'Content-Type': 'application/json'};
      if (token != null) {
        headers['Authorization'] = 'Bearer $token';
      }

      final response = await _client.post(
        Uri.parse('$_apiBase/auth/send-verification-code'),
        headers: headers,
        body: jsonEncode({
          'email': email,
          'type': type, // 'register' or 'reset_password'
        }),
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        if (json['success'] == true) {
          final message = json['message'] as String? ?? '验证码已发送';
          return Result.success(message);
        } else {
          final message = json['message'] as String? ?? '发送失败';
          return Result.failure(Exception(message));
        }
      } else {
        final errorMessage = _parseErrorMessage(response.body);
        return Result.failure(Exception(errorMessage));
      }
    } catch (e) {
      Logger.error('发送验证码异常: $e');
      final errorMsg = _getErrorMessage(e);
      return Result.failure(Exception(errorMsg));
    }
  }

  // 忘记密码
  Future<Result<String>> forgotPassword(
    String email,
    String verificationCode,
    String newPassword,
  ) async {
    try {
      final response = await _client.post(
        Uri.parse('$_apiBase/auth/reset-password'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'email': email,
          'verification_code': verificationCode,
          'new_password': newPassword,
        }),
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        if (json['success'] == true) {
          final message = json['message'] as String? ?? '密码重置成功';
          return Result.success(message);
        } else {
          final message = json['message'] as String? ?? '密码重置失败';
          return Result.failure(Exception(message));
        }
      } else {
        final errorMessage = _parseErrorMessage(response.body);
        return Result.failure(Exception(errorMessage));
      }
    } catch (e) {
      Logger.error('忘记密码异常: $e');
      final errorMsg = _getErrorMessage(e);
      return Result.failure(Exception(errorMsg));
    }
  }

  // 获取用户订阅信息
  Future<Result<UserSubscription>> getUserSubscription() async {
    try {
      final token = await getToken();
      if (token == null) {
        return Result.failure(Exception('未登录'));
      }

      final response = await _client.get(
        Uri.parse('$_apiBase/user/subscription'),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        if (json['success'] == true) {
          final data = json['data'] as Map<String, dynamic>;
          final subscription = UserSubscription.fromJson(data);

          // 保存订阅信息
          final prefs = await SharedPreferences.getInstance();
          await prefs.setString(_subscriptionUrlKey, subscription.universalUrl);
          await prefs.setString(_expireTimeKey, subscription.expireTime);
          await prefs.setInt(_maxDevicesKey, subscription.maxDevices);
          await prefs.setInt(_onlineDevicesKey, subscription.onlineDevices);
          await prefs.setBool(_hasSubscriptionKey, true);

          return Result.success(subscription);
        } else {
          final message = json['message'] as String? ?? '获取订阅信息失败';
          return Result.failure(Exception(message));
        }
      } else {
        final errorMessage = _parseErrorMessage(response.body);
        return Result.failure(Exception(errorMessage));
      }
    } catch (e) {
      Logger.error('获取订阅信息异常: $e');
      final errorMsg = _getErrorMessage(e);
      return Result.failure(Exception(errorMsg));
    }
  }

  // 退出登录
  Future<void> logout() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_tokenKey);
    await prefs.remove(_emailKey);
    await prefs.remove(_usernameKey);
    await prefs.remove(_subscriptionUrlKey);
    await prefs.remove(_expireTimeKey);
    await prefs.remove(_hasSubscriptionKey);
    await prefs.remove(_maxDevicesKey);
    await prefs.remove(_onlineDevicesKey);
  }

  // 获取订阅 URL
  Future<String?> getSubscriptionUrl() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_subscriptionUrlKey);
  }

  // 获取到期时间
  Future<String?> getExpireTime() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_expireTimeKey);
  }

  // 获取最大设备数
  Future<int> getMaxDevices() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getInt(_maxDevicesKey) ?? 0;
  }

  // 获取在线设备数
  Future<int> getOnlineDevices() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getInt(_onlineDevicesKey) ?? 0;
  }

  // 解析错误消息
  String _parseErrorMessage(String body) {
    try {
      final json = jsonDecode(body) as Map<String, dynamic>;
      return json['message'] as String? ?? '请求失败，请检查网络连接';
    } catch (e) {
      return '请求失败，请检查网络连接';
    }
  }

  // 获取错误消息
  String _getErrorMessage(dynamic e) {
    final message = e.toString();
    if (message.contains('timeout')) {
      return '连接超时，请检查网络连接';
    } else if (message.contains('SSL') || message.contains('Certificate')) {
      return 'SSL 连接错误，请检查网络设置';
    } else if (message.contains('Failed host lookup') ||
        message.contains('SocketException')) {
      return '无法连接到服务器，请检查网络';
    } else {
      return '操作失败: ${e.toString()}';
    }
  }

  // 获取套餐列表（不需要认证）
  Future<Result<List<Package>>> getPackages() async {
    try {
      final response = await _client.get(
        Uri.parse('$_apiBase/packages/'),
        headers: {'Content-Type': 'application/json'},
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        if (json['success'] == true || json['success'] == null) {
          final data = json['data'] as List<dynamic>? ?? [];
          final packages = data.map((e) => Package.fromJson(e as Map<String, dynamic>)).toList();
          Logger.debug('获取套餐列表成功: ${packages.length} 个套餐');
          return Result.success(packages);
        } else {
          final message = json['message'] as String? ?? '获取套餐列表失败';
          Logger.error('获取套餐列表失败: $message');
          return Result.failure(Exception(message));
        }
      } else {
        final errorMessage = _parseErrorMessage(response.body);
        Logger.error('获取套餐列表失败: ${response.statusCode}, 错误: $errorMessage');
        return Result.failure(Exception(errorMessage));
      }
    } catch (e) {
      Logger.error('获取套餐列表异常: $e');
      final errorMsg = _getErrorMessage(e);
      return Result.failure(Exception(errorMsg));
    }
  }

  // 创建订单
  Future<Result<Order>> createOrder({
    required int packageId,
    String paymentMethod = 'alipay',
    String? couponCode,
    bool useBalance = false,
    double balanceAmount = 0.0,
  }) async {
    try {
      final token = await getToken();
      if (token == null) {
        return Result.failure(Exception('未登录'));
      }

      final body = <String, dynamic>{
        'package_id': packageId,
        'payment_method': paymentMethod,
        if (couponCode != null) 'coupon_code': couponCode,
        'use_balance': useBalance,
        'balance_amount': balanceAmount,
      };

      final response = await _client.post(
        Uri.parse('$_apiBase/orders/'),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
        body: jsonEncode(body),
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        if (json['success'] == true) {
          final data = json['data'] as Map<String, dynamic>;
          final order = Order.fromJson(data);
          Logger.debug('创建订单成功: ${order.orderNo}');
          return Result.success(order);
        } else {
          final message = json['message'] as String? ?? '创建订单失败';
          Logger.error('创建订单失败: $message');
          return Result.failure(Exception(message));
        }
      } else {
        final errorMessage = _parseErrorMessage(response.body);
        Logger.error('创建订单失败: ${response.statusCode}, 错误: $errorMessage');
        return Result.failure(Exception(errorMessage));
      }
    } catch (e) {
      Logger.error('创建订单异常: $e');
      final errorMsg = _getErrorMessage(e);
      return Result.failure(Exception(errorMsg));
    }
  }

  // 获取订单状态
  Future<Result<OrderStatus>> getOrderStatus(String orderNo) async {
    try {
      final token = await getToken();
      if (token == null) {
        return Result.failure(Exception('未登录'));
      }

      final response = await _client.get(
        Uri.parse('$_apiBase/orders/$orderNo/status'),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        if (json['success'] == true) {
          final data = json['data'] as Map<String, dynamic>;
          final status = OrderStatus.fromJson(data);
          return Result.success(status);
        } else {
          final message = json['message'] as String? ?? '获取订单状态失败';
          return Result.failure(Exception(message));
        }
      } else {
        final errorMessage = _parseErrorMessage(response.body);
        return Result.failure(Exception(errorMessage));
      }
    } catch (e) {
      Logger.error('获取订单状态异常: $e');
      final errorMsg = _getErrorMessage(e);
      return Result.failure(Exception(errorMsg));
    }
  }

  void dispose() {
    _client.close();
  }
}

// Result 类型用于处理成功/失败
class Result<T> {
  final T? data;
  final Exception? error;

  Result.success(this.data) : error = null;
  Result.failure(this.error) : data = null;

  bool get isSuccess => error == null;
  bool get isFailure => error != null;

  R when<R>({
    required R Function(T) success,
    required R Function(Exception) failure,
  }) {
    if (isSuccess && data != null) {
      return success(data as T);
    } else {
      return failure(error!);
    }
  }
}

